rust/azure_iot_operations_services/src/azure_device_registry/client.rs

```diff
@@ -1,2 +1,1604 @@
+    ///
+    /// Returns the updated [`Device`] once updated.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidRequestArgument`](ErrorKind::InvalidRequestArgument)
+    /// if timeout is 0 or > `u32::max`.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if:
+    /// - device or inbound endpoint names are invalid.
+    /// - there are any underlying errors from the AIO RPC protocol.
+    pub async fn update_device_plus_endpoint_status(
+        &self,
+        device_name: String,
+        inbound_endpoint_name: String,
+        status: DeviceStatus,
+        timeout: Duration,
+    ) -> Result<Device, Error> {
+        let status_payload = adr_name_gen::UpdateDeviceStatusRequestPayload {
+            device_status_update: status.into(),
+        };
+        let update_device_status_request =
+            adr_name_gen::UpdateDeviceStatusRequestBuilder::default()
+                .payload(status_payload)
+                .map_err(ErrorKind::from)?
+                .topic_tokens(Self::get_topic_tokens(device_name, inbound_endpoint_name))
+                .timeout(timeout)
+                .build()
+                .map_err(ErrorKind::from)?;
+        let response = self
+            .update_device_status_command_invoker
+            .invoke(update_device_status_request)
+            .await
+            .map_err(ErrorKind::from)?;
+        Ok(response.payload.updated_device.into())
+    }
+
+    /// Starts observation of a [`Device`]'s updates from the Azure Device Registry service.
+    ///
+    /// Note: On cleanup, unobserve should always be called so that the service knows to stop sending notifications.
+    ///
+    /// # Arguments
+    /// * `device_name` - The name of the device.
+    /// * `inbound_endpoint_name` - The name of the inbound endpoint.
+    /// * `timeout` - The duration until the client stops waiting for a response to the request, it is rounded up to the nearest second.
+    ///
+    /// Returns the [`DeviceUpdateObservation`] if the observation was started successfully or [`struct@Error`].
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`DuplicateObserve`](ErrorKind::DuplicateObserve)
+    /// if the [`Device`] is already being observed.
+    ///
+    /// [`struct@Error`] of kind [`InvalidRequestArgument`](ErrorKind::InvalidRequestArgument)
+    /// if timeout is 0 or > `u32::max`.
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if:
+    /// - device or inbound endpoint names are invalid.
+    /// - there are any underlying errors from the AIO RPC protocol.
+    ///
+    /// [`struct@Error`] of kind [`ObservationError`](ErrorKind::ObservationError)
+    /// if the observation was not accepted by the service.
+    pub async fn observe_device_update_notifications(
+        &self,
+        device_name: String,
```

Comment:
This would actually be better as a &str because it would reduce allocations by 1, but honestly at this point we probably need to do a full pass specifically on optimizing strings given how many revisions there have been.

Reply:
NA